### PR TITLE
[FEATURE] add _RECURSIVE option to the subpartParams Configuration of the lookup _TABLE instructions (#479)

### PR DESCRIPTION
**Pull request by vzz3**
Applying my patch [eef0c94] to the typo3v9 branch was easy. Fixing some bugs that prevented me from testing the new table based parameter expansion and applying 2 change sets (18.07.19 [98605ee], 22.07.19 [e17ec22]) that were required for automated testing was the harder part.

Pleas have a look especially  at the changes of the methods getConfigurationsForBranch() and getDuplicateRowsIfExist () of the class CrawlerController in commit [91e88fc]. I changed them in in a way that the fulfil the test cases from the TYPO3 v8 (current master) branch , but I'm not sure if this is intended.

### CODE REVIEW

**Comment by tomasnorre** [comment]
Thanks for your effort @vzz3.
I'll look into this beginning of next week I hope.

File: Classes/Command/BuildQueueCommand.php [hunk 1]
```diff
@@ -128,12 +128,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $crawlerController->registerQueueEntriesInternallyOnly = true;
         }
 
-        if ($this->request instanceof \TYPO3\CMS\Extbase\Mvc\Cli\Request) {
-            $pageId = MathUtility::forceIntegerInRange($input->getOption('startpage'), 0);
-        } else {
+        #if ($this->request instanceof \TYPO3\CMS\Extbase\Mvc\Cli\Request) {
```
Comment (tomasnorre) [inline Classes/Command/BuildQueueCommand.php]: That's not how to comment code in PHP :) `//`
But please don't just comment out code if it's not needed anymore remove it.

File: Classes/Controller/CrawlerController.php [hunk 2]
```diff
@@ -657,7 +666,7 @@ public function getUrlsForPageId($pageId)
      * @param $depth
      * @return array
      *
-     * TODO: Write Functional Tests
+     * TODO: Write Functional Tests => done?!?!
```
Comment (tomasnorre) [inline Classes/Controller/CrawlerController.php]: If I remember correctly, it isn't done yet..

File: Classes/Controller/CrawlerController.php [hunk 3]
```diff
@@ -677,18 +686,21 @@ public function getConfigurationsForBranch(int $rootid, $depth)
         }
         /* @var PageTreeView $tree */
         $tree = GeneralUtility::makeInstance(PageTreeView::class);
-        $perms_clause = $this->backendUser->getPagePermsClause(Permission::PAGE_SHOW);
-        $tree->init('AND ' . $perms_clause);
+        $perms_clause = $this->getBackendUser()->getPagePermsClause(Permission::PAGE_SHOW);
+        $tree->init( empty($perms_clause) ? '' : ('AND ' . $perms_clause) );
```
Comment (tomasnorre) [inline Classes/Controller/CrawlerController.php]: 👍

File: Classes/Controller/CrawlerController.php [hunk 4]
```diff
@@ -677,18 +686,21 @@ public function getConfigurationsForBranch(int $rootid, $depth)
...
         foreach ($tree->tree as $node) {
             $pids[] = $node['row']['uid'];
         }
 
         $queryBuilder = $this->getQueryBuilder('tx_crawler_configuration');
 
-        $queryBuilder->getRestrictions()
-            ->removeAll()
-            ->add(GeneralUtility::makeInstance(DeletedRestriction::class));
+        // The pre TYPO3 v9 version use the method BackendUtility::BEenableFields('tx_crawler_configuration')
```
Comment (tomasnorre) [inline Classes/Controller/CrawlerController.php]: If not needed, please remove

**Comment by tomasnorre** [comment]
Added a few comments, but will look into it next week.

**Comment by vzz3** [comment]
Please tell me when you're done with the review, then I like to fix all changes at once.

File: Classes/Controller/CrawlerController.php [hunk 5]
```diff
@@ -810,25 +823,36 @@ public function expandParameters($paramArray, $pid)
                             if ($fieldName === 'uid' || $GLOBALS['TCA'][$subpartParams['_TABLE']]['columns'][$fieldName]) {
                                 $queryBuilder = $this->getQueryBuilder($subpartParams['_TABLE']);
 
+                                if($recursiveDepth > 0) {
+                                    /** @var \TYPO3\CMS\Core\Database\QueryGenerator $queryGenerator */
+                                    $queryGenerator = GeneralUtility::makeInstance(\TYPO3\CMS\Core\Database\QueryGenerator::class);
```
Comment (tomasnorre) [inline Classes/Controller/CrawlerController.php]: Any reason for the not using the BackendUtility::BEgetRootLine() ? Then we would not need the QueryGenerator

File: Classes/Controller/CrawlerController.php [hunk 6]
```diff
@@ -810,25 +823,36 @@ public function expandParameters($paramArray, $pid)
...
+                                if(!empty($addTable)) {
+                                    // TODO: Check if this works as intended!
+                                    $queryBuilder->add('from', $addTable);
+                                }
                                 $transOrigPointerField = $GLOBALS['TCA'][$subpartParams['_TABLE']]['ctrl']['transOrigPointerField'];
 
                                 if ($subpartParams['_ENABLELANG'] && $transOrigPointerField) {
                                     $queryBuilder->andWhere(
                                         $queryBuilder->expr()->lte(
-                                            $queryBuilder->quoteIdentifier($transOrigPointerField),
```
Comment (tomasnorre) [inline Classes/Controller/CrawlerController.php]: Why is this removed ?

File: Classes/Controller/CrawlerController.php [hunk 7]
```diff
@@ -838,7 +862,7 @@ public function expandParameters($paramArray, $pid)
 
                                 $rows = [];
                                 while ($row = $statement->fetch()) {
-                                    $rows[$fieldName] = $row;
+                                    $rows[$row[$fieldName]] = $row;
```
Comment (tomasnorre) [inline Classes/Controller/CrawlerController.php]: Looks like something, not related to your change, should be refactored here, as we have switch to a domain/model

File: Classes/Controller/CrawlerController.php [hunk 8]
```diff
@@ -1229,13 +1253,15 @@ protected function getDuplicateRowsIfExist($tstamp, $fieldArray)
                 );
         }
 
-        $statement = $queryBuilder
-            ->andWhere('exec_time != 0')
-            ->andWhere('process_id != 0')
+        $queryBuilder
+            /* ->andWhere('exec_time != 0') # this was ' AND NOT exec_time' therfore functional test fails => change back */ ->andWhere('NOT exec_time')
```
Comment (tomasnorre) [inline Classes/Controller/CrawlerController.php]: Remove if not needed.

File: Tests/Functional/Controller/CrawlerControllerTest.php [hunk 9]
```diff
@@ -182,6 +192,300 @@ public function getLogEntriesForSetId($setId, $filter, $doFlush, $doFullFlush, $
...
+
+        unset($configuration['paramExpanded'], $configuration['URLs']);
+        $newCheckSum = md5(serialize($configuration));
+        $this->assertEquals(
+            $newCheckSum,
+            $this->subject->_call('getConfigurationHash', $configuration)
+        );
+    }
+
+// CrawlerController::isCrawlingProtocolHttps() removed in [7f7f07b]
```
Comment (tomasnorre) [inline Tests/Functional/Controller/CrawlerControllerTest.php]: Please remove if not needed

**Comment by tomasnorre** [comment]
You can do you adjustments now. Thanks

File: Classes/Controller/CrawlerController.php (hunk 5 above)
Comment (vzz3) [reply Classes/Controller/CrawlerController.php]: BackendUtility::BEgetRootLine() does a different thing. It returns the path from the provided PID back to the root page (PID=0). However, we need all PIDs of a sub tree.

File: Classes/Controller/CrawlerController.php (hunk 6 above)
Comment (vzz3) [reply Classes/Controller/CrawlerController.php]: because it result in a wrong SQL query. The column name gets "double" quoted.
e.g.:
```suggestion
 [ Doctrine\DBAL\Exception\InvalidFieldNameException ]                                                                           
  An exception occurred while executing 'SELECT `uid` FROM `tx_news_domain_model_news` WHERE (`pid` IN (?, ?, ?)) AND (hidden =   
  0) AND (```l10n_parent``` <= 0) AND (`tx_news_domain_model_news`.`deleted` = 0)' with params [46, 274, 275]:                    
                                                                                                                                  
  Unknown column '`l10n_parent`' in 'where clause'
```

File: Classes/Controller/CrawlerController.php (hunk 7 above)
Comment (vzz3) [reply Classes/Controller/CrawlerController.php]: The _TABLE based parameter expansion doesn't work at all with the old line
```
$rows[$fieldName] = $row;
```
because the array $rows can only contain a single row. The array item $fieldName gets overwritten in ever loop iteration. Therefore, I was forced to fix it, otherwise it would not have bin possible to test any of my changes.
Probably it is possible to just use
```
$rows[] = $row;
```
instead of 
```
$rows[$row[$fieldName]] = $row;
```
. However, my current code produces the exact same result (array) as the code of the crawler master branch. Which hopefully creates as least as possible side effects.
